libevmasm/RuleList.h

```diff
@@ -118,50 +124,51 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart2(
-		{{Instruction::DIV, {X, 0}}, [=]{ return u256(0); }, true},
-		{{Instruction::DIV, {0, X}}, [=]{ return u256(0); }, true},
-		{{Instruction::DIV, {X, 1}}, [=]{ return X; }, false},
-		{{Instruction::SDIV, {X, 0}}, [=]{ return u256(0); }, true},
-		{{Instruction::SDIV, {0, X}}, [=]{ return u256(0); }, true},
-		{{Instruction::SDIV, {X, 1}}, [=]{ return X; }, false},
-		{{Instruction::AND, {X, ~u256(0)}}, [=]{ return X; }, false},
-		{{Instruction::AND, {~u256(0), X}}, [=]{ return X; }, false},
-		{{Instruction::AND, {X, 0}}, [=]{ return u256(0); }, true},
-		{{Instruction::AND, {0, X}}, [=]{ return u256(0); }, true},
-		{{Instruction::OR, {X, 0}}, [=]{ return X; }, false},
-		{{Instruction::OR, {0, X}}, [=]{ return X; }, false},
-		{{Instruction::OR, {X, ~u256(0)}}, [=]{ return ~u256(0); }, true},
-		{{Instruction::OR, {~u256(0), X}}, [=]{ return ~u256(0); }, true},
-		{{Instruction::XOR, {X, 0}}, [=]{ return X; }, false},
-		{{Instruction::XOR, {0, X}}, [=]{ return X; }, false},
-		{{Instruction::MOD, {X, 0}}, [=]{ return u256(0); }, true},
-		{{Instruction::MOD, {0, X}}, [=]{ return u256(0); }, true},
-		{{Instruction::EQ, {X, 0}}, [=]() -> Pattern { return {Instruction::ISZERO, {X}}; }, false },
-		{{Instruction::EQ, {0, X}}, [=]() -> Pattern { return {Instruction::ISZERO, {X}}; }, false },
-		{{Instruction::SHL, {0, X}}, [=]{ return X; }, false},
-		{{Instruction::SHR, {0, X}}, [=]{ return X; }, false},
-		{{Instruction::SHL, {X, 0}}, [=]{ return u256(0); }, true},
-		{{Instruction::SHR, {X, 0}}, [=]{ return u256(0); }, true},
-		{{Instruction::GT, {X, 0}}, [=]() -> Pattern { return {Instruction::ISZERO, {{Instruction::ISZERO, {X}}}}; }, false},
-		{{Instruction::LT, {0, X}}, [=]() -> Pattern { return {Instruction::ISZERO, {{Instruction::ISZERO, {X}}}}; }, false},
-		{{Instruction::GT, {X, ~u256(0)}}, [=]{ return u256(0); }, true},
-		{{Instruction::LT, {~u256(0), X}}, [=]{ return u256(0); }, true},
-		{{Instruction::GT, {0, X}}, [=]{ return u256(0); }, true},
-		{{Instruction::LT, {X, 0}}, [=]{ return u256(0); }, true},
-		{{Instruction::AND, {{Instruction::BYTE, {X, Y}}, {u256(0xff)}}}, [=]() -> Pattern { return {Instruction::BYTE, {X, Y}}; }, false},
-		{{Instruction::BYTE, {31, X}}, [=]() -> Pattern { return {Instruction::AND, {X, u256(0xff)}}; }, false}
+		{{Pattern::Builtins::ADD, {X, 0}}, [=]{ return X; }, false},
+		{{Pattern::Builtins::ADD, {0, X}}, [=]{ return X; }, false},
+		{{Pattern::Builtins::SUB, {X, 0}}, [=]{ return X; }, false},
+		{{Pattern::Builtins::SUB, {~Word(0), X}}, [=]() -> Pattern { return {Pattern::Builtins::NOT, {X}}; }, false},
+		{{Pattern::Builtins::MUL, {X, 0}}, [=]{ return Word(0); }, true},
+		{{Pattern::Builtins::MUL, {0, X}}, [=]{ return Word(0); }, true},
+		{{Pattern::Builtins::MUL, {X, 1}}, [=]{ return X; }, false},
+		{{Pattern::Builtins::MUL, {1, X}}, [=]{ return X; }, false},
+		{{Pattern::Builtins::MUL, {X, Word(-1)}}, [=]() -> Pattern { return {Pattern::Builtins::SUB, {0, X}}; }, false},
```

Comment:
```suggestion
		{{Pattern::Builtins::MUL, {X, ~Word(0)}}, [=]() -> Pattern { return {Pattern::Builtins::SUB, {0, X}}; }, false},
```
(while we're at it anyways: I think that's better)

Reply:
Do you think `Word(-1)` might fail in some way? I wouldn't use `~` in an arithmetic context.

Reply:
Well, I'd have more confidence that ``~Word(0)`` is well defined and actually has all bits set for every word type on which it's defined - whereas that ``Word(-1)`` does the correct thing feels like the greater assumption about the type ``Word`` - but I'd be fine with keeping it at ``Word(-1)`` here, so consider these suggestions "optional".

Reply:
Think
```
include <iostream>

class UWord128 {
public:
    UWord128(unsigned int x) { std::cout << x << std::endl; /* ... */ }
    // no constructor for signed int
};

int main() {
    UWord128 x(-1); // will end up with 2^32-1
}
```

Reply:
But then you are assuming we are using two's complement. What about an assert that `Word(-1) == ~Word(0)`?

Reply:
> But then you are assuming we are using two's complement. What about an assert that `Word(-1) == ~Word(0)`?

I'd make "two's complement" a hard requirement on the type... but yeah - such an assertion would also do the trick I guess.

It'd be great to have a precise list compiling the exact operations needed on the word type and their required semantics - and then maybe even add assertions for all of that at the beginning of the rule list construction - maybe we should add an issue for that (way too much effort for right now I guess, and maybe a bit overkill in general...).

Reply:
Examples for another such assertion we could eventually add would be ``(~Word(0)) + Word(1) == Word(0)`` and ``(~Word(0)) + Word(2) == Word(1)`` to indicate modulo arithmetic, etc...

Reply:
In C++20 we might even be able to put all this into a concept :-D. Given the operations on the word type be constexpr.